From 6266ceedbfaf435866455e9f9242dba3d9ae7309 Mon Sep 17 00:00:00 2001
From: Hyein Yu <hyeinyu@google.com>
Date: Wed, 15 Jan 2025 01:35:39 +0000
Subject: [PATCH] RESTRICT AUTOMERGE Update getDefaultCBRPackageName for fixing
 security vulnerability

Bug: 381419370
Test: atest CellBroadcastServiceTests
Test: Install fake cellbroadcast app. Check if cellbroadcast message is
received
Flag: EXEMPT bug fix

Change-Id: If2e797dd6d352627ddfac16f3d0447be0e172d25
(cherry picked from commit f24f40e2512f450b6a23536a4209318cd3043d99)
Merged-In: If2e797dd6d352627ddfac16f3d0447be0e172d25
---
 .../cellbroadcastservice/CellBroadcastHandler.java     |  6 ++++--
 .../tests/CellBroadcastHandlerTest.java                | 10 ++++++++++
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/src/com/android/cellbroadcastservice/CellBroadcastHandler.java b/src/com/android/cellbroadcastservice/CellBroadcastHandler.java
index 550bd73..4be8ab5 100644
--- a/src/com/android/cellbroadcastservice/CellBroadcastHandler.java
+++ b/src/com/android/cellbroadcastservice/CellBroadcastHandler.java
@@ -855,9 +855,11 @@ public class CellBroadcastHandler extends WakeLockStateMachine {
      * Find the name of the default CBR package. The criteria is that it belongs to CB apex and
      * handles the given intent.
      */
-    static String getDefaultCBRPackageName(Context context, Intent intent) {
+    @VisibleForTesting
+    public static String getDefaultCBRPackageName(Context context, Intent intent) {
         PackageManager packageManager = context.getPackageManager();
-        List<ResolveInfo> cbrPackages = packageManager.queryBroadcastReceivers(intent, 0);
+        List<ResolveInfo> cbrPackages = packageManager.queryBroadcastReceivers(intent,
+                PackageManager.MATCH_SYSTEM_ONLY);
 
         // remove apps that don't live in the CellBroadcast apex
         cbrPackages.removeIf(info ->
diff --git a/tests/src/com/android/cellbroadcastservice/tests/CellBroadcastHandlerTest.java b/tests/src/com/android/cellbroadcastservice/tests/CellBroadcastHandlerTest.java
index bf8b661..6deedc3 100644
--- a/tests/src/com/android/cellbroadcastservice/tests/CellBroadcastHandlerTest.java
+++ b/tests/src/com/android/cellbroadcastservice/tests/CellBroadcastHandlerTest.java
@@ -36,6 +36,7 @@ import android.content.Context;
 import android.content.IIntentSender;
 import android.content.Intent;
 import android.content.res.Resources;
+import android.content.pm.PackageManager;
 import android.database.Cursor;
 import android.database.MatrixCursor;
 import android.net.Uri;
@@ -389,6 +390,15 @@ public class CellBroadcastHandlerTest extends CellBroadcastServiceTestBase {
         cellBroadcastHandler.cleanup();
     }
 
+    @Test
+    @SmallTest
+    public void testGetDefaultCBRPackageName() {
+        Intent intent = new Intent(Telephony.Sms.Intents.ACTION_SMS_EMERGENCY_CB_RECEIVED);
+        CellBroadcastHandler.getDefaultCBRPackageName(mMockedContext, intent);
+        verify(mMockedPackageManager, times(1))
+                .queryBroadcastReceivers(intent, PackageManager.MATCH_SYSTEM_ONLY);
+    }
+
     /**
      * Makes injecting a mock factory easy.
      */
-- 
2.50.1.351.gcf940e82a1-goog

